crates/data_components/src/databricks/auth.rs

```diff
@@ -0,0 +1,192 @@
+/*
+Copyright 2024-2025 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use async_trait::async_trait;
+use secrecy::{ExposeSecret, SecretString};
+use serde::{Deserialize, Serialize};
+use snafu::prelude::*;
+use std::time::Duration;
+use std::{fmt, sync::Arc};
+use tokio::{sync::watch, task::JoinHandle, time::sleep};
+use util::fibonacci_backoff::FibonacciBackoffBuilder;
+
+use crate::token_provider::{Result, TokenProvider};
+
+const TOKEN_REFRESH_BUFFER_SECS: u64 = 300;
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display(
+        "Failed to obtain Databricks service principal token for machine-to-machine authentication."
+    ))]
+    UnableToGetToken { source: Box<dyn std::error::Error> },
+}
+
+#[derive(Clone)]
+pub struct DatabricksM2MTokenProvider {
+    endpoint: String,
+    client_id: String,
+
+    tx: watch::Sender<String>,
+    rx: watch::Receiver<String>,
+
+    _handle: Arc<JoinHandle<()>>,
+}
+
+impl fmt::Debug for DatabricksM2MTokenProvider {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("M2mTokenProvider")
+            .field("endpoint", &self.endpoint)
+            .field("client_id", &self.client_id)
+            .field("tx", &"<watch::Sender>")
+            .field("rx", &"<watch::Receiver>")
+            .field("_handle", &"<JoinHandle>")
+            .finish()
+    }
+}
+
+impl DatabricksM2MTokenProvider {
+    pub async fn try_new(
+        endpoint: String,
+        client_id: String,
+        client_secret: SecretString,
+    ) -> Result<Self, Error> {
+        // initial fetch
+        let TokenResponse {
+            access_token,
+            expires_in,
+            ..
+        } = get_m2m_access_token(endpoint.clone(), client_id.clone(), client_secret.clone())
+            .await
+            .map_err(|e| Error::UnableToGetToken { source: e })?;
+
+        // create watch channel
+        let (tx, rx) = watch::channel(access_token.clone());
+
+        // spawn background refresh loop
+        let cloned_client_id = client_id.clone();
+        let cloned_endpoint = endpoint.clone();
+        let cloned_tx = tx.clone();
+
+        let secret = client_secret.clone();
+
+        let handle = tokio::spawn(async move {
```

Comment:
Would it make sense to use CancellableTaskHandle here? https://github.com/spiceai/spiceai/blob/trunk/crates/runtime/src/cancellable_task.rs

Reply:
Yes, will do in a separate PR